src/anonymeter/evaluators/singling_out_evaluator.py

```diff
@@ -2,129 +2,168 @@
 # Copyright (c) 2022 Anonos IP LLC.
 # See https://github.com/statice/anonymeter/blob/main/LICENSE.md for details.
 """Privacy evaluator that measures the singling out risk."""
+
 import logging
-from typing import Any, Callable, Dict, List, Optional, Set, Tuple
+import operator
+from functools import reduce
+from typing import Any, Callable, Dict, List, Optional, Sequence, Set, Tuple
 
 import numpy as np
 import numpy.typing as npt
 import pandas as pd
-from pandas.api.types import is_bool_dtype, is_numeric_dtype
+import polars as pl
 from scipy.optimize import curve_fit
 
 from anonymeter.stats.confidence import EvaluationResults, PrivacyRisk
 
-rng = np.random.default_rng()
+rng_default = np.random.default_rng()
 logger = logging.getLogger(__name__)
 
 
 def _escape_quotes(string: str) -> str:
     return string.replace('"', '\\"').replace("'", "\\'")
 
 
-def _query_expression(col: str, val: Any, dtype: np.dtype) -> str:
-    """Generate type-aware query expression."""
-    query: str = ""
-
-    if pd.api.types.is_datetime64_any_dtype(dtype):
-        f"{col} == '{val}'"
-    elif isinstance(val, str):
-        query = f"{col} == '{_escape_quotes(val)}'"
-    else:
-        query = f"{col} == {val}"
+def _query_from_record(
+    record: dict,  # or pl.Series, but dict is often easier for single rows
```

Comment:
If `record` can be either `dict` or `pl.Series`, why not typing it accordingly?

```suggestion
    record: Union[dict, pl.Series],  # dict is often easier for single rows
```